### PR TITLE
Support for cubic

### PR DESCRIPTION
**Pull request by Aperence**
See #402  for the details.

This PR provide support for cubic, and support for selecting the maximum datagram size.

Exemple of usage : 
```py
from aioquic.quic.congestion.cubic import CubicCongestionControl
from aioquic.quic.congestion.reno import RenoCongestionControl
from aioquic.quic.congestion.slow_starts import HyStart

if (args.cca == "reno"):
    cca = RenoCongestionControl(max_datagram_size=K_MAX_DATAGRAM_SIZE, callback=callback, slow_start=HyStart(max_datagram_size=K_MAX_DATAGRAM_SIZE))
elif (args.cca == "cubic"):
    cca = CubicCongestionControl(max_datagram_size=K_MAX_DATAGRAM_SIZE, callback=callback, reno_friendly_activated=args.reno_friendly)
else:
    exit(1)

conf = QuicConfiguration(alpn_protocols=["quic-testing"], is_client=True, congestion_control=cca)
```

### CODE REVIEW

File: src/aioquic/quic/configuration.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 from dataclasses import dataclass, field
 from os import PathLike
 from re import split
-from typing import Any, List, Optional, TextIO, Union
+from typing import Any, List, Optional, TextIO, Union, Dict
```
Comment (jlaine) [inline src/aioquic/quic/configuration.py:4]: `Dict` is not used

File: src/aioquic/quic/logger.py [hunk 2]
```diff
@@ -32,6 +32,20 @@
 def hexdump(data: bytes) -> str:
     return binascii.hexlify(data).decode("ascii")
 
+def is_builtin_class(Object):
+    return Object.__class__.__module__ == 'builtins'
+
+def get_dataclass_attr(Object) -> dict:
```
Comment (jlaine) [inline src/aioquic/quic/logger.py:38]: Why is this needed?

**Comment by jlaine** [comment]
This PR still seems to include unrelated code such as changes to slowstart, minmax.. can we cut it down to the bare minimum to support Cubic please?

**Comment by Aperence** [comment]
Hi, 

First, thanks for your feedback.

I have normally removed all content that weren't related to cubic (as slow-start, minmax, get_dataclass_attr, ...).

Only the code for cubic should remain now.

Best,

File: src/aioquic/quic/logger.py (hunk 2 above)
Comment (Aperence) [reply src/aioquic/quic/logger.py]: It was needed for a simpler logging of the BBR/BBRv2 congestion control, where many attributes were present, it was thus better to have some method that automatically gets all its attributes.

This won't be needed for cubic, so I'll simply remove it for now

File: src/aioquic/quic/congestion/congestion.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        self._max_datagram_size = max_datagram_size
+        self.cwnd = fixed_cwnd
+        self.data_in_flight = 0
+
+    def set_recovery(self, recovery):
+        # recovery is a QuicPacketRecovery instance
+        self.recovery = recovery
+
+    def on_init(self, *args, **kwargs):
+        pass
```
Comment (jlaine) [inline src/aioquic/quic/congestion/congestion.py:93]: This is unused AFAICT?

File: src/aioquic/quic/congestion/cubic.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+K_CUBIC_MAX_IDLE_TIME = 2   # reset the cwnd after 2 seconds of inactivity
+
+class CubicCongestionControl(QuicCongestionControl):
+    """
+    Cubic congestion control implementation for aioquic
+    """
+
+    def __init__(self, max_datagram_size : int, callback=None, reno_friendly_activated = True) -> None:
+        super().__init__(max_datagram_size=max_datagram_size, callback=callback)
+        global K_CUBIC_ADDITIVE_INCREASE
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:19]: Uhoh, this doesn't look good, this is de-factor not a constant

File: src/aioquic/quic/congestion/congestion.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    self._increases += 1
+                    if self._increases >= self._size:
+                        return True
+                elif delta > 0:
+                    self._increases = 0
+        return False
+    
+
+def Now():
+    return datetime.timestamp(datetime.now())
```
Comment (jlaine) [inline src/aioquic/quic/congestion/congestion.py:76]: I'm not sure I'm comfortable with this, until now all timestamps have been provided as input to the congestion controller's method. 

If I remember correctly, this timestamp is handed to us by the event loop, and there is no guarantee its aligns with `datetime.now`.

**Comment by Aperence** [comment]
Hello, 

I modified the content you pinpointed above, feel free to still give me feedback is anything is still weird or should be changed :D

Have a nice day,

File: examples/http3_client.py [hunk 6]
```diff
@@ -545,6 +550,14 @@ async def main(
...
     if args.secrets_log:
         configuration.secrets_log_file = open(args.secrets_log, "a")
+    if args.congestion_control:
+        if args.congestion_control == "cubic":
+            configuration.congestion_control = CubicCongestionControl()
+        elif args.congestion_control == "reno":
+            configuration.congestion_control = RenoCongestionControl()
+        else:
+            print("Invalid congestion control algorithm")
+            exit(127)
```
Comment (jlaine) [inline examples/http3_client.py:560]:
```suggestion
            raise Exception("Invalid congestion control algorithm")
```

File: src/aioquic/quic/congestion/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from .congestion import *
```
[the remark this answers is not in the document]
Comment (Aperence) [reply src/aioquic/quic/congestion/__init__.py:1]: This feels indeed a better way to do it

**Comment by jlaine** [comment]
OK we're getting close to a mergeable state! Would you mind rebasing on top of `main` and squashing down all your commits so we can get a CI run?

**Comment by Aperence** [comment]
Normally this should be done, I squashed all the commits into one (4cf275a34e0c8fa30c29b9e9441f0adb31e4015a)

File: src/aioquic/quic/congestion/__init__.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
+        if self.callback:
+            self.callback(CongestionEvent.RTT_MEASURED, self)
+
+    def get_congestion_window(self) -> int:
+        # return the cwnd in number of bytes
+        return self.cwnd   
+    
+    def _set_congestion_window(self, value):
+        self.cwnd = value
```
Comment (jlaine) [inline src/aioquic/quic/congestion/__init__.py:126]: I'm having trouble understanding why we have these accessors?

File: src/aioquic/quic/congestion/__init__.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                # don't call this if it is a superclass that runs
+                self.data_in_flight -= packet.sent_bytes
+
+    def on_packets_lost(self, packets: Iterable[QuicSentPacket], now: float) -> None:
+        if self.callback:
+            self.callback(CongestionEvent.PACKET_LOST, self)
+        if type(self) == QuicCongestionControl:
+            for packet in packets:
+                # don't call this if it is a superclass that runs
+                self.data_in_flight -= packet.sent_bytes
```
Comment (jlaine) [inline src/aioquic/quic/congestion/__init__.py:115]: This looks very odd..

File: src/aioquic/quic/congestion/__init__.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        return self.cwnd   
+    
+    def _set_congestion_window(self, value):
+        self.cwnd = value
+
+    def get_ssthresh(self) -> Optional[int]: 
+        pass
+
+    def get_bytes_in_flight(self) -> int:
+        return self.data_in_flight
```
Comment (jlaine) [inline src/aioquic/quic/congestion/__init__.py:132]: Same as above, why do we need this accessor, especially since `data_in_flight` doesn't seem to be set anywhere other than at initialisation?

File: src/aioquic/quic/congestion/__init__.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+    def log_callback(self) -> Dict[str, Any]:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/__init__.py:134]: This doesn't look like it's actually a "callback", it looks more like `get_log_data` or something?

File: src/aioquic/quic/congestion/__init__.py (hunk 8 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/__init__.py:126]: It is used for having an uniform access to the congestion window, as the variable names might change according to the implementation and the developer, thus by providing a method that must be implemented on each class extending QuicCongestionControll, we ensure that whatever the variable name is, we can still access the congestion window (Note however that the setter is unused, so I'll just remove it)

File: src/aioquic/quic/congestion/__init__.py (hunk 9 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/__init__.py:115]: It is an artifact of how bytes in flight was used, I'll refactor this

File: src/aioquic/quic/congestion/__init__.py (hunk 10 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/__init__.py:121]: Same as above, to provide an uniform interface for getting the bytes in flight, this is needed by `recovery.py` at line 134, which is used typically in `connection.py`, line 580 (this was already there before, just changed it to create a getter)

File: src/aioquic/quic/congestion/__init__.py (hunk 11 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/__init__.py]: It is indeed used to get the logging data for the congestion controller, thus I renamed it

**Comment by jlaine** [comment]
I've been reading more of the code and I'm not entirely comfortable with how it's architectured. Could you take a look at PR #432 which *only* moves code around and introduces a base class for congestion control algorithms?

**Comment by rmarx** [comment]
Hey both,

(just for the QLOG stuff ;))

qlog is actually intended to be extensible, so implementations can log whatever extra data they care about. We indeed don't want to over-specify for the different CC algorithms, but other implementations have certainly also logged their own fields for that. The one thing you can't count on of course is that general tools (like QVIS) will do anything more with your custom fields than just show them (which is what qvis does). 

Looking at the specific fields here, they seem sensible. My suggestion would be to start the fields with `cubic-` though, and make them lowercase (that IS required by qlog ;)), so `cubic-phase` or `cubic-wmax`.

File: src/aioquic/quic/congestion/cubic.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
+        self.now = now
+        super().on_rtt_measurement(latest_rtt, now)
+        # check whether we should exit slow start
+        if self.ssthresh is None and self._rtt_monitor.is_rtt_increasing(
+            latest_rtt, now
+        ):
+            self._cwnd_prior = self.congestion_window
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:181]: This code does nothing, `self._cwnd_prior` is never used anywhere?

File: src/aioquic/quic/congestion/cubic.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        self._starting_congestion_avoidance = False
+        self.K = 0
+        self._W_est = 0
+        self._cwnd_epoch = 0
+        self._t_epoch = 0
+        self._W_max = self.congestion_window
+
+    def on_packet_acked(self, packet: QuicSentPacket, now: float) -> None:
+        super().on_packet_acked(packet, now)
+        rtt = self.recovery._rtt_smoothed
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:72]: No, the congestion control should not have a handle to the `QuicPacketRecovery`. Is it actually the smooth RTT we want here or the latest?

File: src/aioquic/quic/congestion/cubic.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        self._cwnd_epoch = None
+        self._t_epoch = None
+        self._W_max = None
+        self._first_slow_start = True
+        self._starting_congestion_avoidance = False
+        self.K = 0
+        self._W_est = 0
+        self._cwnd_epoch = 0
+        self._t_epoch = 0
+        self._W_max = self.congestion_window
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:68]: There are multiple redundant initialisations of the same variables..

File: src/aioquic/quic/congestion/cubic.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        return target_segments * self._max_datagram_size
+    
+    def is_reno_friendly(self, t) -> bool:
+        return self.reno_friendly_activated and self.W_cubic(t) < self._W_est
+    
+    def is_concave(self):
+        return self.congestion_window < self._W_max
+    
+    def is_convex(self):
+        return self.congestion_window >= self._W_max
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:53]: This is not used, it's just a negation of `is_concave`

File: src/aioquic/quic/congestion/cubic.py (hunk 13 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/cubic.py:76]: It should be according to the RFC the weighted average RTT

File: src/aioquic/quic/congestion/cubic.py (hunk 14 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/cubic.py:72]: Fixed

File: src/aioquic/quic/congestion/cubic.py (hunk 13 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/cubic.py:69]: But I'm wondering if we couldn't use the latest RTT indeed ?

File: examples/http3_client.py [hunk 16]
```diff
@@ -518,6 +518,9 @@ async def main(
     parser.add_argument(
         "--zero-rtt", action="store_true", help="try to send requests using 0-RTT"
     )
+    parser.add_argument(
+        "--congestion-control", type=str, help="which congestion control algorithm to use (reno, cubic)"
+    )
```
Comment (jlaine) [inline examples/http3_client.py:523]: Not needed, already in `main`

File: examples/http3_server.py [hunk 17]
```diff
@@ -564,6 +564,9 @@ async def main(
     parser.add_argument(
         "-v", "--verbose", action="store_true", help="increase logging verbosity"
     )
+    parser.add_argument(
+        "--congestion-control", type=str, help="which congestion control algorithm to use (reno, cubic)"
+    )
```
Comment (jlaine) [inline examples/http3_server.py:569]: Not needed, already in `main`

File: src/aioquic/quic/congestion/base.py [hunk 18]
```diff
@@ -17,15 +17,16 @@ class QuicCongestionControl(abc.ABC):
     congestion_window: int = 0
     ssthresh: Optional[int] = None
 
-    def __init__(self, *, max_datagram_size: int) -> None:
+    def __init__(self, *, max_datagram_size: int, recovery) -> None:
         self.congestion_window = K_INITIAL_WINDOW * max_datagram_size
+        self.recovery = recovery
```
Comment (jlaine) [inline src/aioquic/quic/congestion/base.py:22]: No, let's not have a reference loop please

File: src/aioquic/quic/congestion/base.py [hunk 19]
```diff
@@ -17,15 +17,16 @@ class QuicCongestionControl(abc.ABC):
...
 
     @abc.abstractmethod
-    def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
+    def on_packet_acked(self, *, now : float, packet: QuicSentPacket) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/base.py:25]: Why do you need `now`, you already have `packet.sent_time`?

File: src/aioquic/quic/congestion/base.py [hunk 20]
```diff
@@ -17,15 +17,16 @@ class QuicCongestionControl(abc.ABC):
...
         ...  # pragma: no cover
 
     @abc.abstractmethod
-    def on_packet_sent(self, *, packet: QuicSentPacket) -> None:
+    def on_packet_sent(self, *, now : float, packet: QuicSentPacket) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/base.py:29]: Why do you need `now`, you already have `packet.sent_time`?

File: src/aioquic/quic/congestion/cubic.py [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+        self.last_ack = None
+
+        self.now = 0
+
+    def better_cube_root(self, x):
+        if (x < 0):
+            # avoid precision errors that make the cube root returns an imaginary number
+            return -((-x)**(1./3.))
+        else:
+            return (x)**(1./3.)
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:42]: Move this to a standalone function, it will avoid re-defining it in `test_cubic.py`

File: src/aioquic/quic/congestion/cubic.py (hunk 13 above)
Comment (jlaine) [reply src/aioquic/quic/congestion/cubic.py:69]: I don't know the RFCs enough, if you need it add a `smoothed_rtt` keyword argument to `on_rtt_measurement`

File: src/aioquic/quic/congestion/base.py (hunk 18 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/base.py:22]: How should I get the smoothed RTT in another way ?

File: src/aioquic/quic/recovery.py [hunk 22]
```diff
@@ -62,7 +63,10 @@ def update_bucket(self, now: float) -> None:
             self.evaluation_time = now
 
     def update_rate(self, congestion_window: int, smoothed_rtt: float) -> None:
-        pacing_rate = congestion_window / max(smoothed_rtt, K_MICRO_SECOND)
+        if self.pacing_rate == None or self.pacing_rate == 0:
+            pacing_rate = congestion_window / max(smoothed_rtt, K_MICRO_SECOND)
+        else:
+            pacing_rate = self.pacing_rate
```
Comment (jlaine) [inline src/aioquic/quic/recovery.py:69]: Why are we changing this?

File: src/aioquic/quic/recovery.py [hunk 23]
```diff
@@ -115,17 +119,17 @@ def __init__(
...
         self._cc = create_congestion_control(
-            congestion_control_algorithm, max_datagram_size=max_datagram_size
+            congestion_control_algorithm, max_datagram_size=max_datagram_size, recovery=self
         )
         self._pacer = QuicPacketPacer(max_datagram_size=max_datagram_size)
 
     @property
     def bytes_in_flight(self) -> int:
-        return self._cc.bytes_in_flight
+        return int(self._cc.bytes_in_flight)
```
Comment (jlaine) [inline src/aioquic/quic/recovery.py:128]: This is already an int, no need to cast it.

File: src/aioquic/quic/congestion/base.py (hunk 18 above)
Comment (Aperence) [reply src/aioquic/quic/congestion/base.py]: Changed to latest RTT instead

File: src/aioquic/quic/congestion/base.py [hunk 24]
```diff
@@ -1,5 +1,5 @@
 import abc
-from typing import Dict, Iterable, Optional, Protocol
+from typing import Dict, Iterable, Optional, Protocol, Any
```
Comment (jlaine) [inline src/aioquic/quic/congestion/base.py:2]: Please run the linters (black / ruff / mypy) locally to avoid CI runs,  this is guaranteed to not be OK as it's not sorted

File: src/aioquic/quic/congestion/cubic.py [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+            flight_size = self.bytes_in_flight
+            new_ssthresh = max(int(flight_size*K_CUBIC_LOSS_REDUCTION_FACTOR), K_MINIMUM_WINDOW * self._max_datagram_size)
+            self.ssthresh = new_ssthresh
+            self.congestion_window = max(self.ssthresh, K_MINIMUM_WINDOW * self._max_datagram_size)
+            
+
+            self._starting_congestion_avoidance = True  # restart a new congestion avoidance phase
+
+
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:172]: `rtt` not `latest_rtt`

File: src/aioquic/quic/congestion/cubic.py [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+
+        self._rtt_monitor = QuicRttMonitor()
+
+        self.rtt = 0.02    # starting RTT is considered to be 20ms
+
+        self.reset()
+
+        self.last_ack = None
+
+        self.now = 0
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:44]: Please remove this, the controller cannot know the "current time" as it has no clock. From what I can see the only place you use `self.now` is in `get_log_data`, you can pass it in there.

File: src/aioquic/quic/recovery.py (hunk 22 above)
Comment (Aperence) [reply src/aioquic/quic/recovery.py]: Woopsie, seems like an artifact of BBR (where the congestion controller needed to set the pacing rate), just removed it

**Comment by jlaine** [comment]
Please revert your changes to `reno.py`, the only change should be the addition of `now` in `on_packet_acked`.

File: src/aioquic/quic/connection.py [hunk 27]
```diff
@@ -3318,4 +3319,4 @@ def _write_streams_blocked_frame(
                     is_unidirectional=frame_type == QuicFrameType.STREAMS_BLOCKED_UNI,
                     limit=limit,
                 )
-            )
+            )
```
Comment (jlaine) [inline src/aioquic/quic/connection.py:3322]: Please revert all changes to `connection.py`, they are not needed

File: src/aioquic/quic/logger.py [hunk 28]
```diff
@@ -32,7 +32,6 @@
 def hexdump(data: bytes) -> str:
     return binascii.hexlify(data).decode("ascii")
 
-
```
Comment (jlaine) [inline src/aioquic/quic/logger.py:35]: Please revert all changes to `logger.py`, they are not needed

File: src/aioquic/quic/recovery.py [hunk 29]
```diff
@@ -1,8 +1,9 @@
 import logging
 import math
 from typing import Any, Callable, Dict, Iterable, List, Optional
+from datetime import datetime
```
Comment (jlaine) [inline src/aioquic/quic/recovery.py:4]: This is sure to fail linting, use `black` and `ruff --fix` locally to fix everything in `src`

File: src/aioquic/quic/recovery.py [hunk 30]
```diff
@@ -238,7 +238,7 @@ def on_ack_received(
                 )
 
             # inform congestion controller
-            self._cc.on_rtt_measurement(now=now, rtt=latest_rtt)
+            self._cc.on_rtt_measurement(now=now, latest_rtt=latest_rtt)
```
Comment (jlaine) [inline src/aioquic/quic/recovery.py:241]: Revert this

File: src/aioquic/quic/recovery.py [hunk 31]
```diff
@@ -263,7 +263,7 @@ def on_loss_detection_timeout(self, *, now: float) -> None:
             self._pto_count += 1
             self.reschedule_data(now=now)
 
-    def on_packet_sent(self, *, packet: QuicSentPacket, space: QuicPacketSpace) -> None:
+    def on_packet_sent(self, packet: QuicSentPacket, space: QuicPacketSpace) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/recovery.py:266]: Revert this

File: tests/test_recovery.py [hunk 32]
```diff
@@ -1,7 +1,20 @@
 from unittest import TestCase
 
+<<<<<<< HEAD
```
Comment (jlaine) [inline tests/test_recovery.py:3]: Revert all these changes

File: src/aioquic/quic/congestion/cubic.py [hunk 33]
```diff
@@ -0,0 +1,206 @@
...
+from ..packet_builder import QuicSentPacket
+from typing import Iterable, Dict, Any
+
+# cubic specific variables (see https://www.rfc-editor.org/rfc/rfc9438.html#name-definitions)
+K_CUBIC_K = 1    
+K_CUBIC_C = 0.4
+K_CUBIC_LOSS_REDUCTION_FACTOR = 0.7
+K_CUBIC_MAX_IDLE_TIME = 2   # reset the cwnd after 2 seconds of inactivity
+
+def better_cube_root(self, x):
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:17]: self? While we are at it, please add type annotations otherwise `mypy` will fail

File: src/aioquic/quic/congestion/reno.py [hunk 34]
```diff
@@ -41,14 +41,14 @@ def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
                 self._congestion_stash -= count * self.congestion_window
                 self.congestion_window += count * self._max_datagram_size
 
-    def on_packet_sent(self, *, packet: QuicSentPacket) -> None:
+    def on_packet_sent(self, packet: QuicSentPacket) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/reno.py:44]: The switch to keyword-only arguments was on purpose, revert this

File: src/aioquic/quic/congestion/reno.py [hunk 35]
```diff
@@ -41,14 +41,14 @@ def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
...
         self.bytes_in_flight += packet.sent_bytes
 
-    def on_packets_expired(self, *, packets: Iterable[QuicSentPacket]) -> None:
+    def on_packets_expired(self, packets: Iterable[QuicSentPacket]) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/reno.py:47]: Ditto

File: src/aioquic/quic/congestion/reno.py [hunk 36]
```diff
@@ -41,14 +41,14 @@ def on_packet_acked(self, *, packet: QuicSentPacket) -> None:
...
         for packet in packets:
             self.bytes_in_flight -= packet.sent_bytes
 
-    def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> None:
+    def on_packets_lost(self, packets: Iterable[QuicSentPacket], now: float) -> None:
```
Comment (jlaine) [inline src/aioquic/quic/congestion/reno.py:51]: Ditto

File: src/aioquic/quic/congestion/cubic.py [hunk 37]
```diff
@@ -0,0 +1,206 @@
...
+    K_INITIAL_WINDOW,
+    QuicCongestionControl,
+    QuicRttMonitor,
+    register_congestion_control,
+)
+from ..packet_builder import QuicSentPacket
+from typing import Iterable, Dict, Any
+
+# cubic specific variables (see https://www.rfc-editor.org/rfc/rfc9438.html#name-definitions)
+K_CUBIC_K = 1
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:12]: Where is `K_CUBIC_K` used?

File: src/aioquic/quic/recovery.py (hunk 23 above)
Comment (jlaine) [reply src/aioquic/quic/recovery.py:124]: > In cubic, this can be a float, leading to some unexpected error, so I just casted it when trying to access it

Uh no I don't see how this can be a float, we start with 0 and all we do is add / subtract packet lengths. This cannot and must not be a float.

File: src/aioquic/quic/congestion/cubic.py [hunk 38]
```diff
@@ -0,0 +1,206 @@
...
+
+            t = now - self._t_epoch
+            
+            target = None
+            if (self.W_cubic(t + self.rtt) < self.congestion_window):
+                target = self.congestion_window
+            elif (self.W_cubic(t + self.rtt) > 1.5*self.congestion_window):
+                target = self.congestion_window*1.5
+            else:
+                target = self.W_cubic(t + self.rtt)
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:114]: No need to re-calculate this 3 times, just assign `target = self.W_cubic(t + self.rtt)` then clamp it

File: src/aioquic/quic/congestion/reno.py [hunk 39]
```diff
@@ -66,12 +66,22 @@ def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> N
 
         # TODO : collapse congestion window if persistent congestion
 
-    def on_rtt_measurement(self, *, now: float, rtt: float) -> None:
+    def on_rtt_measurement(self, latest_rtt: float, now: float) -> None:
         # check whether we should exit slow start
         if self.ssthresh is None and self._rtt_monitor.is_rtt_increasing(
-            now=now, rtt=rtt
+            rtt=latest_rtt, now=now
```
Comment (jlaine) [inline src/aioquic/quic/congestion/reno.py:72]: Revert this too

File: src/aioquic/quic/congestion/reno.py [hunk 40]
```diff
@@ -66,12 +66,22 @@ def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> N
...
         ):
             self.ssthresh = self.congestion_window
-
+    
+    def get_log_data(self) -> Dict[str, Any]:
+        data = {
+            "cwnd" : self.congestion_window
+        }
+        if (self.ssthresh == None):
```
Comment (jlaine) [inline src/aioquic/quic/congestion/reno.py:80]:
```suggestion
        if self.ssthresh is None:
```

**Comment by Aperence** [comment]
Ok, I've reverted normally every changed except for cubic, and logs, I've also run the lint checks locally (ruff + black + mypy, thank you btw because I didn't know about these tools at all) and I fixed normally everything that was listed locally

**Comment by Aperence** [comment]
My black version locally seem however to differ compared to the one used on CI, as it fails here, I can fix this (removing a blank line), if needed, but I'm waiting for you to confirm as I might have abused you CI instances with all these push, my apologies for this :cry:

**Comment by jlaine** [comment]
It looks as though you still have one linter error left :)

The code is also going to need thorough tests to check everything is working in all scenarios. I would suggest you try writing a `test_recovery_cubic.py` based on `test_recovery_reno.py`. Full test coverage is a requirement to merge :)

File: src/aioquic/quic/congestion/base.py [hunk 41]
```diff
@@ -40,6 +40,12 @@ def on_packets_lost(self, *, now: float, packets: Iterable[QuicSentPacket]) -> N
     def on_rtt_measurement(self, *, now: float, rtt: float) -> None:
         ...  # pragma: no cover
 
+    def get_log_data(self) -> Dict[str, Any]:
+        data = {"cwnd": self.congestion_window}
```
Comment (jlaine) [inline src/aioquic/quic/congestion/base.py:44]: Looks like you dropped `bytes_in_flight` here

File: src/aioquic/quic/recovery.py [hunk 42]
```diff
@@ -334,10 +334,7 @@ def _get_loss_space(self) -> Optional[QuicPacketSpace]:
         return loss_space
 
     def _log_metrics_updated(self, log_rtt=False) -> None:
-        data: Dict[str, Any] = {
-            "bytes_in_flight": self._cc.bytes_in_flight,
-            "cwnd": self._cc.congestion_window,
-        }
+        data: Dict[str, Any] = self._cc.get_log_data()
         if self._cc.ssthresh is not None:
```
Comment (jlaine) [inline src/aioquic/quic/recovery.py:338]: This doesn't look needed, we're already reporting `ssthresh` in `QuicCongestionControl.get_log_data()`

File: src/aioquic/quic/congestion/cubic.py [hunk 43]
```diff
@@ -0,0 +1,216 @@
...
+    def is_concave(self) -> bool:
+        return self.congestion_window < self._W_max
+
+    def reset(self) -> None:
+        self.congestion_window = K_INITIAL_WINDOW * self._max_datagram_size
+        self.ssthresh = None
+
+        self._first_slow_start = True
+        self._starting_congestion_avoidance = False
+        self.K: float = 0.0
```
Comment (jlaine) [inline src/aioquic/quic/congestion/cubic.py:63]: If you initialise K with a float you don't need the type definition

**Comment by Aperence** [comment]
Normally, I've fixed these last issues + added a test for cubic based on `test_recovery_cubic.py`. I'll try to check later the code that wasn't covered by thos tests

**Comment by jlaine** [comment]
> Normally, I've fixed these last issues + added a test for cubic based on `test_recovery_cubic.py`. I'll try to check later the code that wasn't covered by thos tests

Sure! You can get coverage reports locally too:

```shell
pip install coverage
coverage run -m unittest tests/test_recovery_cubic.py
coverage html
firefox htmlcov/index.html
```

**Comment by jlaine** [comment]
Wow, amazing work, thanks so much for your patience! Is the `test_cubic.py` still useful or is it superseded by your `test_recovery_cubic.py`? If it's still useful maybe merge it into `test_recovery_cubic.py`?

Disclaimer: I am not able to determine whether the code works as intended, or follows specs. Would you mind taking the code in its current form for a spin and checking it still behaves as you want it to?

Once this is done I'd be happy to merge the code!

**Comment by Aperence** [comment]
Hi, 

First, thank you for all your feedback during this PR, I learned quite a lot about CI tools to ensure some code quality !
Secondly, about the `test_cubic.py`, it is still useful, so I'll just merge it in `test_recovery_cubic.py` as you recommended, this should centralize all the tests for the cubic congestion controller.

The implementation seems to still work fine, here are some examples on how it performs:

With no delay on localhost:
image

With a delay of 100ms on localhost:
image

We can clearly still see the cubic growth when the delay is quite important (second plot), and the reno friendly region when we have a reasonable delay (first plot), thus I think we can affirm that it still work as intended

**Comment by jlaine** [comment]
Alright, thanks for the checks! Are these graphs produced from the QLOG output? I saw an article which mentioned you had set up a test bench as part of your University work?

If wouldn't mind squashing your commits down to a single commit with a clean commit message I'd be happy to hit the "merge" button.

**Comment by Aperence** [comment]
Yup, these graphs are generated based on the QLOG files, the script can be found [here](https://github.com/Aperence/linfo1341-congestion-student/blob/main/src/plot_congestion.py) if you are interested with it (it support normally any congestion control, as it only check the presence of the cwnd field in the logs), but some features (W_max, ...) might not be up to date with the latest modifications for the logs (but the core for plotting losses and cwnd does). You can also select a particular field if you want to plot other things (for example ssthresh, bytes_in_flight, ...)

Normally, the commits should be squashed also !

**Comment by jlaine** [comment]
Merged, thanks a lot! 🎉

I hope we will have the pleasure of seing more of your work :)
